homeassistant/components/decora_wifi/light.py

```diff
@@ -19,84 +15,73 @@
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
 
+from . import DecoraWifiAsyncClient
+from .const import DOMAIN
+
 _LOGGER = logging.getLogger(__name__)
 
 # Validation of the user's configuration
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
     {vol.Required(CONF_USERNAME): cv.string, vol.Required(CONF_PASSWORD): cv.string}
 )
 
-NOTIFICATION_ID = "leviton_notification"
-NOTIFICATION_TITLE = "myLeviton Decora Setup"
-
 
-def setup_platform(
+async def async_setup_platform(
     hass: HomeAssistant,
     config: ConfigType,
     add_entities: AddEntitiesCallback,
     discovery_info: DiscoveryInfoType | None = None,
 ) -> None:
-    """Set up the Decora WiFi platform."""
+    """Set up decora_wifi component."""
 
-    email = config[CONF_USERNAME]
-    password = config[CONF_PASSWORD]
-    session = DecoraWiFiSession()
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=config
+        )
+    )
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Decora WiFi platform."""
 
+    session: DecoraWiFiSession = hass.data[DOMAIN][entry.entry_id]
+    asyncSession = DecoraWifiAsyncClient(session, hass)
     try:
-        success = session.login(email, password)
-
-        # If login failed, notify user.
-        if success is None:
-            msg = "Failed to log into myLeviton Services. Check credentials."
-            _LOGGER.error(msg)
-            persistent_notification.create(
-                hass, msg, title=NOTIFICATION_TITLE, notification_id=NOTIFICATION_ID
-            )
-            return
-
-        # Gather all the available devices...
-        perms = session.user.get_residential_permissions()
-        all_switches = []
-        for permission in perms:
-            if permission.residentialAccountId is not None:
-                acct = ResidentialAccount(session, permission.residentialAccountId)
-                for residence in acct.get_residences():
-                    for switch in residence.get_iot_switches():
-                        all_switches.append(switch)
-            elif permission.residenceId is not None:
-                residence = Residence(session, permission.residenceId)
-                for switch in residence.get_iot_switches():
-                    all_switches.append(switch)
-
-        add_entities(DecoraWifiLight(sw) for sw in all_switches)
+        permissions = await asyncSession.get_permissions()
+        residences = await asyncSession.get_residences(permissions)
+        iot_switches = await asyncSession.get_iot_switches(residences)
+        async_add_entities([DecoraWifiLight(e) for e in iot_switches])
+
+    # As of the current release of the decora wifi lib (1.4), all api errors raise a generic ValueError
```

Comment:
When does this happen? Is this a flaky API? If this currently happens we just raise an error, log it and the entry is "loaded". If this only happens when something is wrong, I would say raise ConfigEntryNotReady as it will retry a bit later

Reply:
makes sense, fixed.
unfortunately current latest release lib raises ValueError for all exceptions. This has been fixed on master ([see](https://github.com/tlyakhov/python-decora_wifi/blob/master/decora_wifi/__init__.py#L105)) but not released yet. I started an issue. I think raising `ConfigEntryNotReady` might be best we can do.

Reply:
That's fine with me!